### ISSUE TITLE
Testing multiple workers

### ISSUE THREAD

**Issue by Tomas0413** [channelcat/sanic#804]
Hello, I'm trying out sanic with multiple workers and not getting the results, I'd expect.

This is my server side code:
```
from sanic import Sanic
from sanic.response import json
from sanic.response import text
import time

app = Sanic(__name__)

@app.route("/")
async def test(request):
   print("something")
   time.sleep(5)
   return json("OK")

app.run(host="0.0.0.0", port=8000, workers=20)
```

And this is my client side code:
```
import curio
import curio_http

async def fetch_one(url):
   async with curio_http.ClientSession() as session:
       response = await session.get(url)
       content = await response.json()
       return response, content


async def main(url_list):
   tasks = []

   for url in url_list:
       task = await curio.spawn(fetch_one(url))
       tasks.append(task)

   for task in tasks:
       response, content = await task.join()

       print('GET %s' % response.url)
       print(content)
       print()


url_list = [
   'http://127.0.0.1:8000/',
   'http://127.0.0.1:8000/',
   'http://127.0.0.1:8000/',
   'http://127.0.0.1:8000/',
   'http://127.0.0.1:8000/'
]

if __name__ == '__main__':
   curio.run(main(url_list))
```

When I run with workers=8, this is what I see.

Server:
```
# python3 sanic_server.py 
2017-06-18 22:25:45 - (sanic)[DEBUG]: 
                 ▄▄▄▄▄
        ▀▀▀██████▄▄▄       _______________
      ▄▄▄▄▄  █████████▄  /                 \
     ▀▀▀▀█████▌ ▀▐▄ ▀▐█ |   Gotta go fast!  |
   ▀▀█████▄▄ ▀██████▄██ | _________________/
   ▀▄▄▄▄▄  ▀▀█▄▀█════█▀ |/
        ▀▀▀▄  ▀▀███ ▀       ▄▄
     ▄███▀▀██▄████████▄ ▄▀▀▀▀▀▀█▌
   ██▀▄▄▄██▀▄███▀ ▀▀████      ▄██
▄▀▀▀▄██▄▀▀▌████▒▒▒▒▒▒███     ▌▄▄▀
▌    ▐▀████▐███▒▒▒▒▒▐██▌
▀▄▄▄▄▀   ▀▀████▒▒▒▒▄██▀
          ▀▀█████████▀
        ▄▄██▀██████▀█
      ▄██▀     ▀▀▀  █
     ▄█             ▐▌
 ▄▄▄▄█▌              ▀█▄▄▄▄▀▀▄
▌     ▐                ▀▀▄▄▄▀
 ▀▀▄▄▀

2017-06-18 22:25:45 - (sanic)[INFO]: Goin' Fast @ http://0.0.0.0:8000
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10308]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10310]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10309]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10311]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10313]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10312]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10314]
2017-06-18 22:25:45 - (sanic)[INFO]: Starting worker [10315]
something
something
something
2017-06-18 22:25:56 - (network)[INFO][127.0.0.1:43728]: GET http://127.0.0.1/  200 4
2017-06-18 22:25:56 - (network)[INFO][127.0.0.1:43726]: GET http://127.0.0.1/  200 4
something
2017-06-18 22:25:56 - (network)[INFO][127.0.0.1:43730]: GET http://127.0.0.1/  200 4
2017-06-18 22:26:01 - (network)[INFO][127.0.0.1:43734]: GET http://127.0.0.1/  200 4
something
2017-06-18 22:26:06 - (network)[INFO][127.0.0.1:43732]: GET http://127.0.0.1/  200 4
```

Client:

```
# time python3 sanic_client.py 
GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK


real	0m15.117s
user	0m0.100s
sys	0m0.008s

```

When I run with workers=20, this is what I see.

Server:

```
# python3 sanic_server.py 
2017-06-18 22:28:09 - (sanic)[DEBUG]: 
                 ▄▄▄▄▄
        ▀▀▀██████▄▄▄       _______________
      ▄▄▄▄▄  █████████▄  /                 \
     ▀▀▀▀█████▌ ▀▐▄ ▀▐█ |   Gotta go fast!  |
   ▀▀█████▄▄ ▀██████▄██ | _________________/
   ▀▄▄▄▄▄  ▀▀█▄▀█════█▀ |/
        ▀▀▀▄  ▀▀███ ▀       ▄▄
     ▄███▀▀██▄████████▄ ▄▀▀▀▀▀▀█▌
   ██▀▄▄▄██▀▄███▀ ▀▀████      ▄██
▄▀▀▀▄██▄▀▀▌████▒▒▒▒▒▒███     ▌▄▄▀
▌    ▐▀████▐███▒▒▒▒▒▐██▌
▀▄▄▄▄▀   ▀▀████▒▒▒▒▄██▀
          ▀▀█████████▀
        ▄▄██▀██████▀█
      ▄██▀     ▀▀▀  █
     ▄█             ▐▌
 ▄▄▄▄█▌              ▀█▄▄▄▄▀▀▄
▌     ▐                ▀▀▄▄▄▀
 ▀▀▄▄▀

2017-06-18 22:28:09 - (sanic)[INFO]: Goin' Fast @ http://0.0.0.0:8000
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10828]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10829]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10827]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10826]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10831]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10833]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10835]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10830]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10837]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10832]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10841]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10840]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10839]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10834]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10838]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10842]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10836]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10843]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10844]
2017-06-18 22:28:09 - (sanic)[INFO]: Starting worker [10845]
something
something
something
something
2017-06-18 22:28:20 - (network)[INFO][127.0.0.1:43862]: GET http://127.0.0.1/  200 4
2017-06-18 22:28:20 - (network)[INFO][127.0.0.1:43854]: GET http://127.0.0.1/  200 4
2017-06-18 22:28:20 - (network)[INFO][127.0.0.1:43858]: GET http://127.0.0.1/  200 4
2017-06-18 22:28:20 - (network)[INFO][127.0.0.1:43856]: GET http://127.0.0.1/  200 4
something
2017-06-18 22:28:25 - (network)[INFO][127.0.0.1:43860]: GET http://127.0.0.1/  200 4
```

Client:
```
# time python3 sanic_client.py 
GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK

GET 127.0.0.1:8000/
OK


real	0m10.113s
user	0m0.096s
sys	0m0.012s

```

I'd expect if workers >=5 when my client (which does HTTP GET to 5 URLs in parallel) would finish in 5+ seconds.

Any comments are appreciated!

**Comment by Tomas0413** [comment]
Few more tests:

workers=200

All 5 requests are executed in parallel:

```
real	0m5.111s
user	0m0.100s
sys	0m0.008s
```

workers=100
4 requests finish first and then the 5th one:

```
real	0m10.115s
user	0m0.100s
sys	0m0.008s

```

**Comment by Tomas0413** [comment]
```
# pip3 show sanic
Name: sanic
Version: 0.5.4
Summary: A microframework based on uvloop, httptools, and learnings of flask
Home-page: http://github.com/channelcat/sanic/
Author: Channel Cat
Author-email: channelcat@gmail.com
License: MIT
Location: /usr/local/lib/python3.5/dist-packages
Requires: httptools, ujson, uvloop, aiofiles, websockets
```

**Comment by Tomas0413** [comment]
@r0fls Thanks.  time.sleep(5) is just one example, in fact, what I want to be able to do is to open multiple HTTP connections send the data back to the client. 

For example:
```
@app.route("/")
async def test(request):
   print("something")
   response = urlopen('https://www.bbc.co.uk')
   print(response)
   return json("OK")
```

Do you know why when the worker number is very high all threads get executed at once, but then worker number is lower, not all threads get executed at once?

**Comment by stopspazzing** [comment]
`2017-06-18 22:25:45 - (sanic)[DEBUG]: ` meaning debug is on, but your code shows no debug enabled code. So am I missing something? 
Have you tried with debug OFF? just curious..

**Comment by mczp** [comment]
@Tomas0413 Unless urlopen is an async function (which you should await for) sanic will not bring you any benefit of performance over sync frameworks like flask.
You should either wrap urlopen in a threaded executor (ex. http://pythonhosted.org/asyncio_extras/#module-asyncio_extras.threads) or use an aiohttp based client.
Once done you will not need that many workers to get the perf you need

**Comment by r0fls** [comment]
@Tomas0413 I would use an asynchronous http client for that, like aiohttp. I found that it was far more performant and stable than using the requests library with sanic. Like @mczp said, if you have other components that aren't async compatible then it will slow down the server. I haven't observed the below situation either:
>Do you know why when the worker number is very high all threads get executed at once, but then worker number is lower, not all threads get executed at once?

When I ran your example, it was somewhat random when rerunning the same script, until I used the asyncio sleep, which behaved as expected.
